src/main/scala/scala/tools/refactoring/implementations/Rename.scala

```diff
@@ -122,21 +124,137 @@ abstract class Rename extends MultiStageRefactoring with MarkOccurrences with Tr
         (TextChange(occurence.source, occurence.start, occurence.end, newName), tree)
       }
 
-      val newSourceChanges = textChangesWithTrees.flatMap { case (textChange, tree) =>
+      val renameFileChanges = textChangesWithTrees.flatMap { case (textChange, tree) =>
         if (tree.pos.source.file.name == singleTreeSelection.name + ".scala") {
           Some(RenameSourceFileChange(tree.pos.source.file, newName + ".scala"))
         } else {
           None
         }
       }.distinct
 
+      val moveToDirChanges = {
+        mkMoveToDirChanges(textChangesWithTrees.map(_._2), singleTreeSelection.symbol, newName)
+      }
+
       val textChanges = textChangesWithTrees.map(_._1)
 
-      textChanges ::: newSourceChanges
+      textChanges ::: renameFileChanges ::: moveToDirChanges
     }
 
     trace(s"Old name is ${singleTreeSelection.name}")
     if (singleTreeSelection.name == newName) Right(Nil)
     else Right(generateChanges)
   }
+
+  /*
+   * If we rename a package, it will typically be necessary to move its contents to a new directory.
+   * This method generates these changes for all files where the primary package (see below for exact definition)
+   * corresponds to the directory the files lies in. We are considering a package to be the primary package of a file,
+   * if all non package declarations lie beneath it, and no more deeply nested package satisfies this property. This
+   * is best understood by examples:
+   *
+   * Ex. 1:
+   *     package a.b //<-- primary
+   *
+   *     class C
+   *
+   * Ex. 2:
+   *     package a
+   *     package object b { //<-- primary
+   *       val x = 42
+   *     }
+   *
+   * Ex. 3:
+   *     package a { //<-- primary
+   *       package b {
+   *         class C
+   *       }
+   *       class C
+   *     }
+   */
+  private def mkMoveToDirChanges(changedTrees: List[Tree], selectedSymbol: Symbol, newName: String): List[MoveToDirChange] = {
+    if (!selectedSymbol.hasPackageFlag) {
+      Nil
+    } else {
+      changedTrees.flatMap { tree =>
+        val pkgSymbol = associatedPackageClassSymbol(tree)
+
+        pkgSymbol.flatMap  { pkgSymbol =>
+          index.rootOf(tree).flatMap { root =>
+            val pkgChain = packageChainReflectedInDirLayout(root)
+            val adaptedPkgChain = pkgChain.map {
+              case (dir, sym) if sym == pkgSymbol => (newName, pkgSymbol)
+              case other => other
+            }
+
+            if (adaptedPkgChain == pkgChain) {
+              None
+            } else {
+              val sourceFile = tree.pos.source.file
+              val newPath = {
+                val parent = parentPath(sourceFile, adaptedPkgChain.size + 1)
+                val prefix = if (parent.isEmpty) "" else parent + "/"
+                prefix + adaptedPkgChain.map(_._1).mkString("/")
```

Comment:
I think `/` should be safe, but to find for sure we would have to test on Windows... unfortunately I don't have a Windows box at hand; do you by chance?